[cocoa] Set cookie partition identifier in network resource requests

sysrqb opened the pull request:
#### 428cc313e773a471858ac38d144d4f7ac9379fc7
<pre>
[cocoa] Set cookie partition identifier in network resource requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=280836">https://bugs.webkit.org/show_bug.cgi?id=280836</a>
<a href="https://rdar.apple.com/problem/137220553">rdar://problem/137220553</a>

Reviewed by Sihui Liu.

WebKit, on cocoa platforms, previously supported unconditional partitioned
third-party cookies. That was removed in 203318@main. We are exploring an
implementation of opt-in partitioned third-party cookie (CHIPS), which other
engines are implementing. This is a step towards adding support.

The partition key is the top-level site (i.e., schemeful registrable domain).

New test for partitioned cookies, skipped on all platforms.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/cookies/accept-partitioned-first-and-third-party-cookies-expected.txt: Added.
* LayoutTests/http/tests/cookies/accept-partitioned-first-and-third-party-cookies.html: Added.
* LayoutTests/http/tests/cookies/resources/set-cookie-and-redirect-back.py:
* Source/WebCore/platform/Site.cpp:
(WebCore::Site::string const):
* Source/WebCore/platform/Site.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::cookiePartitionIdentifier):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::updateTaskWithStoragePartitionIdentifier):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
(WebKit::WebSocketTask::WebSocketTask):

Canonical link: <a href="https://commits.webkit.org/286830@main">https://commits.webkit.org/286830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b04545750033f717e4f15331077abb93c56b55a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81749 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79304 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65370 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4518 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60475 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18527 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50436 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66247 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40777 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47839 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23745 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26792 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24080 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83174 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4567 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3077 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68751 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66220 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68007 "Found 2 new API test failures: /TestWebKit:WebKit.AboutBlankLoad, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/test-page-list (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11993 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10083 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11958 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4514 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7329 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4533 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6292 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -251,6 +251,13 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+void NetworkTaskCocoa::updateTaskWithStoragePartitionIdentifier(const WebCore::ResourceRequest& request)
+{
+    // FIXME: Remove respondsToSelector when available with NWLoader. rdar://134913391
+    if (auto* networkStorageSession = m_networkSession->networkStorageSession(); networkStorageSession && networkStorageSession->isOptInCookiePartitioningEnabled() && [task() respondsToSelector:@selector(set_storagePartitionIdentifier:)])
```
Nit: the condition is too complicated to be put in one-line; we could add early return for check failure, e.g. `if (!networkStorageSession) return`.
Also we should use `CheckedPtr`

szewai replied:
If `isOptInCookiePartitioningEnabled` is enabled, but cookie policy is allowing all cookies, do we need to set partition identifier?

sysrqb replied:
Yes, we should still respect the `Partitioned` attribute of the cookie and put it into the correct partition, even if we are accepting all cookies.

szewai replied:
My question is more about whether the request is allowed to use non-partitioned cookies (i.e. search in non-partitioned cookie storage) if the partitionIdentifier is set?

sysrqb replied:
Currently, the CFNetwork implementation will allow access to both partitioned and unpartitioned cookies. I have a follow-up patch that restricts access to only partitioned cookies. I want this patch to focus on setting the partitionIdentifier.

szewai commented on Source/WebCore/platform/Site.h:
```diff
@@ -63,6 +64,7 @@ class Site {
 private:
     String m_protocol;
     RegistrableDomain m_domain;
+    String m_string;
```
You are not initializing or setting `m_string` anywhere so `m_string` is always null and I am not sure why we need this data member. 
`string()` should just return based on the value of `m_protocol` and `m_domain`

sysrqb replied:
I'm setting it in `string()`, it's just lazy initialized. You think we should just make a new string every time?

sysrqb replied:
> string() should just return based on the value of m_protocol and m_domain

I made this change.

szewai commented on Source/WebCore/platform/Site.h:
```diff
@@ -42,6 +42,7 @@ class Site {
 
     const String& protocol() const { return m_protocol; }
     const RegistrableDomain& domain() const { return m_domain; }
+    String string();
```
`String string() const;`

szewai commented on Source/WebCore/platform/Site.cpp:
```diff
@@ -52,4 +52,9 @@ bool Site::matches(const URL& url) const
 
+String Site::string()
+{
+    return makeString(m_protocol, "://"_s, m_domain.string());
```
Should empty site actually mean no partition (i.e. we return `""` here?)

sysrqb replied:
Yes, good catch.

szewai commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -199,6 +200,25 @@ bool NetworkStorageSession::shouldExemptDomainPairFromThirdPartyCookieBlocking(c
+    return switchOn(origin.data(), [&origin] (const SecurityOriginData::Tuple&) -> String {
+        return cookiePartitionIdentifier(origin.toURL());
+    }, [] (const ProcessQualified<OpaqueOriginIdentifier>&) -> String {
+        return "null"_s;
```
We are returning "null" for opaque origin, but "://" for empty site. Do we intentionally treat them as different cases?

sysrqb replied:
I think we want different partitions. I don't see why an opaque origin should get access to non-partitioned cookies. But I'm actually going to delete this function in this patch - it is only needed in a follow-up patch for supporting Clear-Site-Data.

szewai commented on LayoutTests/TestExpectations:
```diff
@@ -3361,6 +3361,7 @@ webkit.org/b/3652 http/tests/misc/prefetch-purpose.html [ Skip ]
 
 # First-party-only cookie policy only supported on Cocoa platforms.
 http/tests/cookies/only-accept-first-party-cookies.html [ Skip ]
+http/tests/cookies/accept-partitioned-first-and-third-party-cookies.html [ Skip ]
```
This seems to skip the test on all platforms. We need a bug (radar) tracking enabling them on Cocoa platforms.

szewai commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -251,6 +251,21 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+{
+    if (auto* networkStorageSession = m_networkSession->networkStorageSession()) {
+        if (!networkStorageSession)
+            return;
```
The condition check should be put outside of `if (auto* networkStorageSession...)`, should be something like:
```
CheckPtr networkStorageSession = m_networkSession->networkStorageSession();
if (!networkStorageSession)
    return;
if (! networkStorageSession->isOptInCookiePartitioningEnabled())
    return;
....
```

Commit:
[cocoa] Set cookie partition identifier in network resource requests
https://bugs.webkit.org/show_bug.cgi?id=280836
rdar://problem/137220553

Reviewed by Sihui Liu.

WebKit, on cocoa platforms, previously supported unconditional partitioned
third-party cookies. That was removed in 203318@main. We are exploring an
implementation of opt-in partitioned third-party cookie (CHIPS), which other
engines are implementing. This is a step towards adding support.

The partition key is the top-level site (i.e., schemeful registrable domain).

New test for partitioned cookies, skipped on all platforms.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/cookies/accept-partitioned-first-and-third-party-cookies-expected.txt: Added.
* LayoutTests/http/tests/cookies/accept-partitioned-first-and-third-party-cookies.html: Added.
* LayoutTests/http/tests/cookies/resources/set-cookie-and-redirect-back.py:
* Source/WebCore/platform/Site.cpp:
(WebCore::Site::string const):
* Source/WebCore/platform/Site.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::cookiePartitionIdentifier):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::updateTaskWithStoragePartitionIdentifier):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
(WebKit::WebSocketTask::WebSocketTask):

Canonical link: https://commits.webkit.org/286830@main
